src/pybind/mgr/dashboard_v2/HACKING.rst

```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+
+To help in the development of the above scenario we added the support for
+asynchronous tasks. To trigger the execution of an asynchronous task we must
+use the follwoing class method of the ``TaskManager`` class::
+
+  import ..tools import TaskManager
+  # ...
+  TaskManager.run(namespace, metadata, func, args, kwargs)
+
+* ``namespace`` is a string that can be used to group tasks. For instance
+  for RBD image creation tasks we could specify ``"rbd/create"`` as the
+  namespace, or conversly ``"rbd/remove"`` for RBD image removal tasks.
+
+* ``metadata`` is a dictionary where we can store key-value pairs that
+  characterize the task. For instance, when creating a task for creating
+  RBD images we can specify the metadata argument as
+  ``{'pool_name': "rbd", image_name': "test-img"}``.
+
+* ``func`` is the python function that implements the operation code, which
+  will be executed asynchronously.
+
+* ``args`` and ``kwargs`` are the positional and named argguments that will be
+  passed to ``func`` when the task manager starts its execution.
+
+The ``TaskManager.run`` method triggers the asynchronous execution of function
+``func`` and returns an ``AsyncTask`` object.
+The ``AsyncTask`` provides the public method ``AsyncTask.wait(timeout)``, which
+can be used to wait for the task to complete up to a timeout defined in seconds
+and provided as an argument. If not argument is provided the ``wait`` method
+blocks until the task is finished.
+
+The ``AsyncTask.wait`` is very useful for tasks that usually are fast
+to execute but that sometimes may take a long time to run.
+The return value of the ``AsyncTask.wait`` method is a pair ``(state, value)``
+where ``state`` is an integer with following possible values:
+
+* ``VALUE_DONE = 0``
+* ``VALUE_EXECUTING = 1``
+* ``VALUE_EXCEPTION = 2``
+
+The ``value`` will store the result of the execution of function ``func`` if
+``state == VALUE_DONE``. If ``state == VALUE_EXECUTING`` then
+``value == None``, and if ``state == VALUE_EXCEPTION`` then ``value`` stores
+the exception object raised by the execution of function ``func``.
+
+The pair ``(namespace, metadata)`` should univocally identify the task being
+run, which means that if you try to trigger a new task that matches the same
+``(namespace, metadata)`` pair of the currently running task, then the new task
+is not created and you get the task object of the current running task.
+
+
+How to get the list of executing and finished asynchronous tasks?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The list of executing and finished tasks is included in the ``Summary``
+controller, which is already polled every 5 seconds by the dashboard frontend.
+But we also provide a dedicated controller to get the same list of executing
+and finished tasks.
+
+The ``Task`` controller exposes the ``/api/task`` endpoint that returns the
+list of executing and finished tasks. This endpoint accepts the ``namespace``
```

Comment:
> list of executing and finished tasks

There are dozens of places where you differentiate between executing and finished tasks. Instead, I'd keep all tasks  in one list and add a `state` property to each task.

Reply:
I think having them separate is easier for the consumers. I believe the executing tasks list will be the most used, but will have less entries than the finished tasks list. Also, I don't think consumers will process both kinds of tasks at the same time, therefore it's more beneficial to have them separate.

But maybe some of the frontend guys can give their opinion on this as they will be the main consumers: @ricardoasmarques @tspmelo @votdev ?

Reply:
I'm especially talking about the backend implementation. The API should be of course designed to be easily consumable by the front end. But keep in mind that the current task queue dialog does accept them as a single list of entries

Reply:
@sebastian-philipp what current task queue dialog are you talking about?

Reply:
@rjfd http://demo.openattic.org/openattic/ one of the icons in the top right corner

Reply:
ATM I see 2 scenarios where this list(s) will be used by the frontend.

1) To display the status of an item, example: "asynchronous RBD deletion"

![screenshot from 2018-03-14 13-11-10](https://user-images.githubusercontent.com/14297426/37404270-4b78e9f8-2789-11e8-874b-aac9d1b65ddb.png)

To implement this scenario , we just need to use the "executing_tasks" list, "finished_tasks" can be ignored here.

2) To display background tasks

![screenshot from 2018-03-14 13-06-42](https://user-images.githubusercontent.com/14297426/37404375-94d9d8be-2789-11e8-8544-1040ac5e35a0.png)

If this prototype (that I'm currently working on) is accepted, having 2 lists is useful.

-----

So, in both scenarios having 2 lists matches the front-end requirements. Of course, if the back-end returns a single list, we can split it in the front-end.